app/demo-interactor/KNDemoKernel.cu

```diff
@@ -0,0 +1,191 @@
+                              Real3*                       direction,
+                              bool*                        alive)
+{
+    // Grid-stride loop, see
+    for (int tid = blockIdx.x * blockDim.x + threadIdx.x;
+         tid < static_cast<int>(states.vars.size());
+         tid += blockDim.x * gridDim.x)
+    {
+        ParticleTrackView particle(params, states, ThreadId(tid));
+        particle = initial_state;
+
+        // Particles begin alive and in the +z direction
+        direction[tid] = {0, 0, 1};
+        alive[tid]     = true;
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform a single interaction.
+ */
+__global__ void iterate_kn(ParticleParamsPointers const         params,
+                           ParticleStatePointers const          states,
+                           KleinNishinaInteractorPointers const kn_params,
+                           SecondaryAllocatorPointers const     secondaries,
+                           RngStatePointers const               rng_states,
+                           Real3* const                         direction,
+                           bool* const                          alive,
+                           real_type* const energy_deposition)
+{
+    SecondaryAllocatorView allocate_secondaries(secondaries);
+
+    for (int tid = blockIdx.x * blockDim.x + threadIdx.x;
+         tid < static_cast<int>(states.vars.size());
+         tid += blockDim.x * gridDim.x)
+    {
+        // Zero energy deposition from this thread, before skipping
+        energy_deposition[tid] = 0;
+
+        // Skip loop if already dead
+        if (!alive[tid])
+        {
+            continue;
+        }
+
+        // Construct particle accessor from immutable and thread-local data
+        ParticleTrackView particle(params, states, ThreadId(tid));
+
+        if (particle.energy() < KleinNishinaInteractor::min_incident_energy())
+        {
+            // Particle is below interaction energy; kill and deposit energy
+            energy_deposition[tid] = particle.energy();
+            alive[tid]             = false;
+            continue;
+        }
+
+        // Construct RNG and interaction interfaces
+        RngEngine              rng(rng_states, ThreadId(tid));
+        KleinNishinaInteractor interact(
+            kn_params, particle, direction[tid], allocate_secondaries);
+
+        // Perform interaction: should emit a single particle (an electron)
+        auto interaction = interact(rng);
+        CHECK(interaction);
+        CHECK(interaction.secondaries.size() == 1);
+
+        // Deposit energy from the secondary (effectively, an infinite energy
+        // cutoff)
+        energy_deposition[tid] = interaction.secondaries.front().energy;
+
+        // Update post-interaction state (apply interaction)
+        direction[tid] = interaction.direction;
+        particle.energy(interaction.energy);
```

Comment:
A side note, with the `auto` in from of interaction and the code pattern, I can not tell from reading the code whether all the result of the "interaction" where applies (i.e. in this case, it would be true if and only if the interaction has only 2 members (secondaries and direction).

Reply:
Good point about `auto`, it can be confusing. I've explicitly changed the return type to Interaction.